### ISSUE TITLE
Impossible to download png on IE/Edge

### ISSUE THREAD

**Issue by adpeyre** [apexcharts/apexcharts.js#518]
Hi,

When I click in the toolbar to download the graph in png, there is no result on IE and Edge.
I have the same issue when I want to use the dataUri() method with these browsers.

Can you help me?

**Comment by adam2872** [comment]
Hi,

Is there any update on this issue, or does anyone have a workaround to allows the downloading of an image of the chart from IE11 or Edge?

Many thanks for an excellent development.

**Comment by d-amit** [comment]
Any update on this issue?

**Comment by vuphuctho** [comment]
Looking for update on this issue.

**Comment by WvanDam** [comment]
Did you check the demo's? 

Using Safari I'm experiencing issues with the download / dataURI function on two of my stacked charts, while the official demo works fine, as well as two other charts.

**Comment by vuphuctho** [comment]
@WvanDam 
https://apexcharts.com/javascript-chart-demos/line-charts/basic/
https://apexcharts.com/javascript-chart-demos/column-charts/basic/
Above are the 2 demos I check in Microsoft Edge. Clicking on "Download PNG" transforms the chart to an image: it does not download the image file and in console I find following warning message: 
```
DOM7001: Invalid argument 'url'. Failed to revoke Blob URL: 
```

**Comment by thibaultchazal** [comment]
Having the same issue and also looking for an update here.
Anyone ?

**Comment by junedchhipa** [comment]
Please try loading this polyfill - https://github.com/eligrey/Blob.js before loading ApexCharts.
The latest version 3.12.0 requires 3 different polyfills to be loaded in order to work correctly with IE11. Those are listed here - https://github.com/apexcharts/apexcharts.js/blob/master/README.md#using-it-with-ie11

Please try loading these polyfills and report here if it works or not.

**Comment by junedchhipa** [comment]
Please upgrade apexcharts to v3.15.1 

Add these IE11 polyfills in the header

```html
<script>
  window.Promise || document.write('<script src="https://cdn.jsdelivr.net/npm/promise-polyfill@8/dist/polyfill.min.js"><\/script>');
  window.Promise || document.write('<script src="https://cdn.jsdelivr.net/npm/eligrey-classlist-js-polyfill@1.2.20171210/classList.min.js"><\/script>');
  window.Promise || document.write('<script src="https://cdn.jsdelivr.net/npm/findindex_polyfill_mdn"><\/script>');
  window.Promise || document.write('<script src="https://cdn.jsdelivr.net/npm/blob-polyfill"><\/script>');		
</script>
```

It should work now.

**Comment by jaksim** [comment]
I've done a bit of digging in IE11 and there are three issues I can see:
1. IE11's svg serialization produces garbage that it then refuses to parse. Issue was raised about this in svg.js repo and it was pretty much resolved as wontfix (https://github.com/svgdotjs/svg.js/issues/298#issuecomment-388312191). It's pretty easy to fix with some regex replacements though. I've prepared this, but then I hit the much worse issue:
2. In IE11, loading svg taints the canvas and so calling canvas.toDataURL() throws SecurityError. The only solution that I'm aware of is to use a library like canvg, which parses the svg and uses canvas methods to redraw it.
3. There's also the issue that neither IE11, nor Edge support navigating to data: urls and so the function triggerDownload will not work anyway. This can be worked around by using canvas.msToBlob and then triggering download of the result using navigator.msSaveBlob instead. This is the sole fix necessary to make the png download work in Edge, but you still need untainted canvas in IE11, otherwise canvas.msToBlob throws SecurityError.

I'm not sure about the revokeObjectURL issue. I mean, I've seen it, but I haven't yet explored the cause.

@junedchhipa Seems like quite a lot of effort. If you don't have any ideas, I'll try to explore the feasibility a bit more soon. For the record, the Blobs polyfill you've linked is not needed.

**Comment by junedchhipa** [comment]
Thanks for the insights, I haven't explored this until now. I definitely appreciate any workarounds you can find.
Yes, I have removed the Blob polyfill from the README.

**Comment by jaksim** [comment]
The rabbit hole goes deeper.

I've got the download working in IE11. However, the success is only partial, since canvg does not (and will not: https://github.com/canvg/canvg/issues/584#issuecomment-413088712) support foreignObject*. Therefore the exported image doesn't contain a legend. I don't really see a way out of this one (except rewriting the way the legend is created to create svg instead of html elements, but I have no idea about the implications of that).

\* There's a little note here, since we're talking about IE11 and exports - the fact that IE11 doesn't support foreignObject also means that the legend is not part of SVG exports either.

I've opened PR https://github.com/apexcharts/apexcharts.js/pull/1296 as a partial fix for now.
